### ISSUE TITLE
Double checking the num_layer assignment in mds_ensemble_postprocessor

### ISSUE THREAD

**Issue by Dengyu-Wu** [Jingkang50/OpenOOD#227]
[line 192](https://github.com/Jingkang50/OpenOOD/blob/18c6f5174a2f518e2a8e819ffb1cd1914bcf12e0/openood/postprocessors/mds_ensemble_postprocessor.py#L192) num_layer = len(feature_type_list), which only ranges from 1-3 according to config file [line 6](https://github.com/Jingkang50/OpenOOD/blob/18c6f5174a2f518e2a8e819ffb1cd1914bcf12e0/configs/postprocessors/mds_ensemble.yml#L6). Is it correct?

Should it be  num_layer = the number of layers or feature?

**Comment by zjysteven** [comment]
This is expected, as in the official implementation, the authors only picked the output of the residual blocks (rather than every single layer).
https://github.com/pokaxpoka/deep_Mahalanobis_detector/blob/90c2105e78c6f76a2801fc4c1cb1b84f4ff9af63/models/resnet.py#L162-L177

**Comment by Dengyu-Wu** [comment]
Thanks for the fast reply. I can understand that only outputs of the residual blocks would be used, but there is not only one residual connection, isn't it? My understanding is that num_layer would be a value related to the number of the features, not process_feature_type. 

Another issue is that I tried 'flat' in [line 6](https://github.com/Jingkang50/OpenOOD/blob/18c6f5174a2f518e2a8e819ffb1cd1914bcf12e0/configs/postprocessors/mds_ensemble.yml#L6). It ran out of 64 GB CPU memory on Cifar10 and crashed the workstation. A fix of this issue would also be helpful.

**Comment by zjysteven** [comment]
Yes your understanding is correct and that is what I tried to explain in my point 2 in the previous comment.

The memory issue is expected and there won't be a "fix" for it because with 'flat' you flattening out the feature map and dealing with a very high-dimensional vector. Especially when the batch size is large, it's expected that it will blow your memory. The original Mahalanobis implementation always takes the mean if I remember correctly, so for example if a feature map is 256 x 16 x 16, it will take the mean over the last two dimensions and get a 256-d vector, which will be used later to compute feature statistics and perform OOD detection (this is also achieved by OpenOOD when using 'mean'). Yet with 'flat', you will be using the whole 256x16x16=65536-d vector as the feature.

**Comment by Dengyu-Wu** [comment]
I see, thanks a lot.
